src/shared/cpu-set-util.c

```diff
@@ -287,3 +295,103 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
         s = (CPUSet) {};
         return 0;
 }
+
+static int numa_policy_to_mempolicy(const NUMAPolicy *policy, unsigned long *maxnode, unsigned long **nodes) {
+        unsigned node, bits, ulong_bits;
+        _cleanup_free_ unsigned long *out = NULL;
+
+        assert(policy);
+        assert(maxnode);
+        assert(nodes);
+
+        /* Take a shortcut here if policy is set to default or local since kernel requires that we pass NULL as nodemask */
+        if (policy->type == NUMA_POLICY_DEFAULT || policy->type == NUMA_POLICY_LOCAL)
+                goto out;
+
+        bits = policy->nodes.allocated * 8;
+        ulong_bits = sizeof(unsigned long) * 8;
+
+        out = new0(unsigned long, (bits / ulong_bits) + (bits % ulong_bits  > 0));
+        if (!out)
+                return -ENOMEM;
+
+        for (node = 0; node < bits; node++)
+                if (CPU_ISSET_S(node, policy->nodes.allocated, policy->nodes.set))
+                        out[node / ulong_bits] |= 1ul << (node % ulong_bits);
+
+out:
+        *nodes = TAKE_PTR(out);
+        *maxnode = num_numa_nodes() + 1;
```

Comment:
I mean in theory it would be fine to use the code as is because it doesn't make sense to set bits in a mask higher than maximum count of NUMA nodes on the box.

Reply:
Let's say I'm on a system with 8192 CPUs, but `policy->nodes.allocated` is 128. Then `*maxnode` would become 8193, but `*nodes*` would point to a buffer of 128 bytes or . The man page says that "nodemask points to a bit mask of node IDs that contains up to maxnode bits". The syscall would then fail with EFAULT.

> Also it appears that first thing that kernel does when it tries to copy bitmask from userspace is that it decrements maxnode.
>
> https://github.com/torvalds/linux/blob/2f4c53349961c8ca480193e47da4d44fdb8335a8/mm/mempolicy.c#L1286

Duh, so it seems. That seems broken or at least contradictory to the man page.